### PR TITLE
[EC-279] DB snapshot based regression test

### PR DESCRIPTION
**Pull request by rtkaczyk**
`Ledger`/`EVM` regression test based on a DB snapshot from an actual syncing client. Goes through all the blocks, replaying the transactions and comparing results.

To run:

```
sbt -Dconfig.file=path/to/config snappy:test
```

An example configuration file can be found at: `src/snappy/resources/example.conf`. Additionally a genesis JSON file may need to be provided (in `blockchain` config section).

### CODE REVIEW

File: src/snappy/resources/example.conf [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  homestead-block-number = "1150000"
+  eip150-block-number = "2500000"
+  eip155-block-number = "3000000"
+  eip160-block-number = "3000000"
+  difficulty-bomb-pause-block-number = "3000000"
+  difficulty-bomb-continue-block-number = "5000000"
+  dao-fork-block-number = "1920000"
+  dao-fork-block-hash = "94365e3a8c0b35089c1d1195081fe7489b528a84b22199c916180db8b28ade7f"
+  account-start-nonce = "0"
+  chain-id = "3d"
```
Comment (AlanVerbner) [inline src/snappy/resources/example.conf]: `0x3d`

File: src/snappy/scala/io/iohk/ethereum/snappy/SnappyTest.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  val pre = new Prerequisites(config)
+  import pre._
+
+  "Blockchain regression test" in {
+
+    val startN = targetBlockchain match {
+      case Some(tb) => config.startBlock.getOrElse(findHighestBlockNumber(tb) - 1).max(1)
+      case None => BigInt(1)
+    }
+    val targetN = config.targetBlock.getOrElse(findHighestBlockNumber(sourceBlockchain)).max(1)
```
Comment (AlanVerbner) [inline src/snappy/scala/io/iohk/ethereum/snappy/SnappyTest.scala]: Minor: What about using `io.iohk.ethereum.db.storage.AppStateStorage#getBestBlockNumber` ?

File: src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala [hunk 3]
```diff
@@ -119,7 +119,7 @@ trait NodeStatusBuilder {
   lazy val nodeStatusHolder = Agent(nodeStatus)
 }
 
-trait BlockChainBuilder {
+trait BlockchainBuilder {
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/nodebuilder/NodeBuilder.scala]: Super Minor: I like naming standardization so probably we will need to change other Block**C**hain names, for example:

- `io.iohk.ethereum.ets.blockchain.BlockChainTestConfig`
- `io.iohk.ethereum.blockchain.sync.FastSync.SyncingHandler#assignBlockChainWork`

WDYT?

File: src/snappy/scala/io/iohk/ethereum/snappy/SnappyTest.scala (hunk 2 above)
Comment (rtkaczyk) [reply src/snappy/scala/io/iohk/ethereum/snappy/SnappyTest.scala]: I would like to hold off with this change. The reason is I would need `sourceAppStateStorage` and `targetAppStateStorage` and it's starting to get messy. One of the changes planned for EC-303 refactor is to have the `Blockchain` be a single point of interaction with blockchain relevant data. That includes obtaining and modifying the best block number. For now I'm just gonna add a TODO comment, ok?

Comment (AlanVerbner) [reply src/snappy/scala/io/iohk/ethereum/snappy/SnappyTest.scala]: Works for me :+1:

File: src/snappy/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+`Ledger`/`EVM` regression test based on a DB snapshot from an actual syncing client. Goes through all the blocks, replaying the transactions and comparing results.
+
+To run:
+
+```
+sbt -Dconfig.file=path/to/config snappy:test
```
Comment (ntallar) [inline src/snappy/README.md]: :+1:

**Pull request merged**